### PR TITLE
Introduces new "env_var" and "file" fields to Secret to allow specifying name/mountPath on injection #minor

### PR DESCRIPTION
**Pull request by gpgn**
# TL;DR
Introduces new fields to the `Secret` object:
- `env_var`
- `file`

Allowing users to directly specify a name or mountPath for the Secret, without having to specify a full PodTemplate(name). The old `mount_requirement` can still be used. Example:

```python

from flytekit import task, workflow, Secret
import flytekit
import os


@task(
    secret_requests=[
        Secret(
            group="user-info",
            key="user_secret",
            mount_requirement=Secret.MountType.FILE,
        ),
        Secret(
            group="user-info",
            key="user_secret",
            mount_requirement=Secret.MountType.ENV_VAR,
        ),
    ]
)
def old() -> None:
    context = flytekit.current_context()
    print("old")

    # mount_requirement=ENV_VAR
    print(context.secrets.get(env_name="_FSEC_USER-INFO_USER_SECRET"))

    # mount_requirement=FILE
    with open('/etc/flyte/secrets/user-info/user_secret', 'r') as infile:
        print(infile.readlines())
    return True

@task(
    secret_requests=[
        Secret(
            group="user-info",
            key="user_secret",
            env_var=Secret.MountEnvVar(name="foo"),
        ),
        Secret(
            group="user-info",
            key="user_secret",
            file=Secret.MountFile(path="/foo"),
        ),
    ]
)

def new() -> None:
    context = flytekit.current_context()
    print("new")

    # env_var=
    print(context.secrets.get(env_name="foo"))

    # file=
    with open('/foo/user_secret', 'r') as infile:
        print(infile.readlines())


@workflow
def training_workflow(hyperparameters: dict) -> None:
    """Put all of the steps together into a single workflow."""
    old()
    new()

```

## Type
 - [ ] Bug Fix
 - [x] Feature
 - [ ] Plugin

## Are all requirements met?

 - [x] Code completed
 - [x] Smoke tested
 - [x] Unit tests added
 - [x] Code documentation added
 - [x] Any pending items have an associated Issue

## Complete description
 _How did you fix the bug, make the feature etc. Link to any design docs etc_

## Tracking Issue
_Remove the '*fixes*' keyword if there will be multiple PRs to fix the linked issue_

fixes https://github.com/flyteorg/flyte/issues/<number>

## Follow-up issue
_NA_
OR
_https://github.com/flyteorg/flyte/issues/<number>_

### CODE REVIEW

File: protos/flyteidl/core/security.proto [hunk 1]
```diff
@@ -43,6 +43,11 @@ message Secret {
     // will depend on the key management system.
     // +optional
     MountType mount_requirement = 4;
+
+    // The name of the environment variable, if the Secret is injected as environment variable. If ommitted, the default
+    // FLYTE_SECRETS_ENV_PREFIX prefix will be used.
+    // +optional
+    string env_name = 5;
```
Comment (EngHabu) [inline protos/flyteidl/core/security.proto]: This only applies to `ENV_VAR` mount type. I would make the change as following:

1. Mark `MountType mount_requirement = 4 [deprecated=true];` as deprecated.
2. Add:
    ```protobuf=
      oneof MountTarget mount_target {
        MountEnvVar env_var = 5;
        MountFile file = 6;
      }

      message MountEnvVar {
        string name = 1;
      }

      message MountFile {
        string path = 1;
      }
    ```
3. Update secrets webhook to look at the new field first, if it's nil, it should fallback to the old field.
4. Update flytekit to always set both fields; `mount_requirement` and `mount_target`

**Review by EngHabu — COMMENTED** [review]
Thank you for your contribution!
I see the original issue is marked as closed as it suggests using PodTemplate. Is that not enough? or not ergonomic enough?

File: protos/flyteidl/core/security.proto (hunk 1 above)
Comment (gpgn) [reply protos/flyteidl/core/security.proto]: Thanks for the feedback! I'll try to add it. Could you explain what you mean with 4? Not sure I get what you mean here.

Comment (gpgn) [reply protos/flyteidl/core/security.proto]: Added the changes with a working initial version, I'll get to adding a number of tests as well.

Comment (gpgn) [reply protos/flyteidl/core/security.proto]: Tests & documentation added, let me know what you think :)

Comment (gpgn) [reply protos/flyteidl/core/security.proto]: I've opened PRs here for the change, I'll close this one.

- https://github.com/flyteorg/flytekit/pull/1726
- https://github.com/flyteorg/flytepropeller/pull/589
- https://github.com/flyteorg/flyteidl/pull/423
- https://github.com/flyteorg/flytesnacks/pull/1017

**Pull request closed without merging**